platform/logging/LoggerFactory.ts

```diff
@@ -0,0 +1,127 @@
+import { LoggingConfig, LoggerConfigType } from './LoggingConfig';
+import { LogLevel } from './LogLevel';
+import { Logger, BaseLogger, LoggerAdapter } from './Logger';
+import { Appenders, DisposableAppender } from './appenders/Appenders';
+import { BufferAppender } from './appenders/buffer/BufferAppender';
+
+/**
+ * Interface that defines a way to retrieve a logger.
+ */
+export interface LoggerFactory {
+  /**
+   * Returns a `Logger` instance for the specified context.
+   * @param context Context to return logger for.
+   */
+  get(context: string): Logger;
+}
+
+/** @internal */
+export class MutableLoggerFactory implements LoggerFactory {
+  private config?: LoggingConfig;
+  private readonly appenders: Map<string, DisposableAppender> = new Map();
+  private readonly bufferAppender = new BufferAppender();
+  private readonly loggers: Map<string, LoggerAdapter> = new Map();
+
+  get(context: string): Logger {
+    if (this.loggers.has(context)) {
+      return this.loggers.get(context)!;
+    }
+
+    let loggerLevel, loggerAppenders;
+    if (this.config) {
+      const { level, appenders } = this.getLoggerConfigByContext(
+        this.config,
+        context
+      )!;
+      loggerLevel = LogLevel.fromId(level);
+      loggerAppenders = appenders.map(
+        appenderKey => this.appenders.get(appenderKey)!
+      );
+    } else {
+      // If we don't have config yet, use `buffered` appender that will store all logged messages in the memory
+      // until the config is ready.
+      loggerLevel = LogLevel.All;
+      loggerAppenders = [this.bufferAppender];
+    }
+
+    this.loggers.set(
+      context,
+      new LoggerAdapter(new BaseLogger(context, loggerLevel, loggerAppenders))
+    );
+
+    return this.loggers.get(context)!;
+  }
+
+  /**
+   * Updates all current active loggers with the new config values.
+   * @param config New config instance.
+   * @returns Promise that is resolved once all loggers are successfully updated.
+   */
+  async updateConfig(config: LoggingConfig) {
+    // Config update is asynchronous and may require some time to complete, so we should invalidate
+    // config so that new loggers will be using BufferAppender until newly configured appenders are ready.
+    // TODO: Before disposing of appenders, we should switch all existing loggers to BufferAppender first.
```

Comment:
The potential problem I was trying to avoid with async/await is when we dispose and create the same file at the same time, I'm not sure how well it will work. Let's say we have FileAppender and config has changed (file path is still the same), if we don't wait for old file appender to close the file before we try to open a new stream what would happen? What do you think?